sys/net/network_layer/ng_ipv6/netif/ng_ipv6_netif.c

```diff
@@ -440,14 +519,172 @@ static ipv6_addr_t *_match_prefix(kernel_pid_t pid, const ipv6_addr_t *addr,
     return NULL;
 }
 
-ipv6_addr_t *ng_ipv6_netif_match_prefix(kernel_pid_t pid, const ipv6_addr_t *prefix)
+/**
+ * @brief Determines the scope of the given address.
+ *
+ * @param[in] addr              The IPv6 address to check.
+ * @param[in] maybe_multicast   False if @p addr is definitely no multicast
+ *                              address, true otherwise.
+ *
+ * @return The scope of the address.
+ *
+ * @pre address is not loopback or unspecified.
+ * see http://tools.ietf.org/html/rfc6724#section-4
+ */
+static uint8_t _get_scope(const ipv6_addr_t *addr, const bool maybe_multicast)
+{
+    if (maybe_multicast && ipv6_addr_is_multicast(addr)) {
+        return (addr->u8[1] & 0x0f);
+    }
+    else if (ipv6_addr_is_link_local(addr)) {
+        return IPV6_ADDR_MCAST_SCP_LINK_LOCAL;
+    }
+    else if (ipv6_addr_is_site_local(addr)) {
+        return IPV6_ADDR_MCAST_SCP_SITE_LOCAL;
+    }
+    else {
+        return IPV6_ADDR_MCAST_SCP_GLOBAL;
+    }
+}
+
+/** @brief Find the best candidate among the configured addresses
+ *          for a certain destination address according to the 8 rules
+ *          specified in RFC 6734, section 5.
+ * @see http://tools.ietf.org/html/rfc6724#section-5
+ *
+ * @param[in] iface              The interface for sending.
+ * @param[in] dest               The destination IPv6 address.
+ * @param[in, out] candidate_set The preselected set of candidate addresses as
+ *                               a bitfield.
+ *
+ * @pre @p dest is not unspecified.
+ *
+ * @return The best matching candidate found on @p iface, may be NULL if none
+ *         is found.
+ */
+static ipv6_addr_t *_source_address_selection(ng_ipv6_netif_t *iface,
+                                      const ipv6_addr_t *dest,
+                                      uint8_t *candidate_set)
 {
-    return _match_prefix(pid, prefix, false);
+    /* create temporary set for assigning "points" to candidates wining in the
+     * corresponding rules.
+     */
+    uint8_t winner_set[NG_IPV6_NETIF_ADDR_NUMOF];
+    memset(winner_set, 0, NG_IPV6_NETIF_ADDR_NUMOF);
+
+    uint8_t max_pts = 0;
+
+    /* _create_candidate_set() assures that `dest` is not unspecified and if
+     * `dest` is loopback rule 1 will fire anyway.  */
```

Comment:
My reasoning was:
Currently, if `dst == ::1` this function is not called, because it's handled as a special case (as you described). If this special handling is gonna be removed at some point, one need to assign `::1` to an (each?) interface and then src = dst would apply.